### PR TITLE
fix(catalog-backend): db query

### PR DESCRIPTION
**Pull request by codermango**
### Background
I am working on the paginated catalog table by using the new queryEntities api, I noticed a issue that the api returns entities which are not belong to the Kind I specify, and also I noticed it also returns some duplicated entities as well. So I looked into the source code, and found some suspicious code which probably related to the issue.

### What I found

After I investigated the source code, I found the following sql code which is generated for fetching the cursor based paginated data.

```sql
select *
from  search  inner join  final_entities 
on  search . entity_id  =  final_entities . entity_id 
where  search . key  = 'metadata.name'
  and ((((( search . entity_id  in (select  search . entity_id  from  search  where  key  = 'kind'
  and ( value  = 'documentation')))))))
  and value 
    > 'something'
   or  ( value  = 'something' -- parentheses here might be wrong
  and  search . entity_id 
    > '70661536-64fd-8466-9640-a5485126a63a')
order by  value  asc,  search . entity_id  asc
limit 11
```

so this part looks wired: 
```sql
and value 
    > 'something'
   or  ( value  = 'something' -- parentheses here might be wrong
  and  search . entity_id 
    > '70661536-64fd-8466-9640-a5485126a63a')
```

I don't fully understand the whole logic of this dbQuery, but I feel the parentheses is placed in a wrong place, it probably should be: 
```sql
and ( value  -- parentheses probably should be here
    > 'something'
   or  value  = 'something'
  and  search . entity_id 
    > '70661536-64fd-8466-9640-a5485126a63a')
```

So the full sql should be: 
```sql
select *
from  search  inner join  final_entities 
on  search . entity_id  =  final_entities . entity_id 
where  search . key  = 'metadata.name'
  and ((((( search . entity_id  in (select  search . entity_id  from  search  where  key  = 'kind'
  and ( value  = 'documentation')))))))
  and ( value  -- parentheses probably should be here
    > 'something'
   or  value  = 'something'
  and  search . entity_id 
    > '70661536-64fd-8466-9640-a5485126a63a')
order by  value  asc,  search . entity_id  asc
limit 11
```

After this fix, at least the issue I found was fixed.

### Conclusion
Please note I don't really understand the whole logic of the dbQuery, so it would be great if you guys can verify and test in your side.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by codermango** [comment]
Hey @vinzscam , thank you for your quick reply and explanation!

Some questions about the case you mentioned: [A, B, C, D, D, E, F, G, H]

1. Why are there duplicated D in the database? Is it supposed to be unique?
2. What was the reason it returns the entities which are not the Kind I specify, is it something related to this as well? Or my database is messed up for some reason?

**Comment by codermango** [comment]
Sure, so I noticed two different issues, **but they seem all happened only when we do cursor request, not the initial request**.

1. We get duplicate entities returned from the api, we only have `default` namespace and the duplicated entities are actually the same entityRef.
The request of this case:
```
/catalog/entities/by-query?cursor=eyJvcmRlckZpZWxkcyI6W3siZmllbGQiOiJtZXRhZGF0YS5uYW1lIiwib3JkZXIiOiJhc2MifV0sImlzUHJldmlvdXMiOmZhbHNlLCJvcmRlckZpZWxkVmFsdWVzIjpbInBsYXRhZm9ybWEiLCIwZmM0NjU1Yy1jY2M1LTQ3ZGUtYTkxMC03ZTYyYTk1ZTgxNzciXSwiZmlsdGVyIjp7ImFueU9mIjpbeyJhbGxPZiI6W3sia2V5Ijoia2luZCIsInZhbHVlcyI6WyJDb21wb25lbnQiXX1dfV19LCJmaXJzdFNvcnRGaWVsZFZhbHVlcyI6WyJhZ2dyaWdhdG8tc2VydmVyIiwiNTVkNjE2YjQtNzQ5Mi00YmY0LTg1YjUtNzIwMjcyODdiMmQ0Il0sInRvdGFsSXRlbXMiOjY3fQ==&limit=10
```
After decode it is:
```json
{
  "orderFields": [{ "field": "metadata.name", "order": "asc" }],
  "isPrevious": false,
  "orderFieldValues": ["entity-name", "0fc4655c-ccc5-47de-a910-7e45412e8177"],
  "filter": {
    "anyOf": [{ "allOf": [{ "key": "kind", "values": ["Component"] }] }]
  },
  "firstSortFieldValues": [
    "another-entity-name",
    "55d616b4-7492-4bf4-85b5-7512478962d4"
  ],
  "totalItems": 67
}
```

2. We get entities that are not belong to the Kind we request, for example we get API entities when we request Documentation entities.
The request is: 
```
/catalog/entities/by-query?cursor=eyJvcmRlckZpZWxkcyI6W3siZmllbGQiOiJtZXRhZGF0YS5uYW1lIiwib3JkZXIiOiJhc2MifV0sImlzUHJldmlvdXMiOmZhbHNlLCJvcmRlckZpZWxkVmFsdWVzIjpbInVwLWxldmVsLW1hbmFnZXItZG9jIiwiN2Y4ZDI1MGUtYjY4MC00YTUwLWJmZDQtNWJkMzhjYjJjMzIzIl0sImZpbHRlciI6eyJhbnlPZiI6W3siYWxsT2YiOlt7ImtleSI6ImtpbmQiLCJ2YWx1ZXMiOlsiRG9jdW1lbnRhdGlvbiJdfV19XX0sImZpcnN0U29ydEZpZWxkVmFsdWVzIjpbImFyZ29jZGRzIiwiYWQ2NTJjOTgtYmYxOC00YTZiLWI4NDgtNDY4NTMzNTUxOWYxIl0sInRvdGFsSXRlbXMiOjgxfQ==&limit=5
```
After decode it is:
```json
{
  "orderFields": [{ "field": "metadata.name", "order": "asc" }],
  "isPrevious": false,
  "orderFieldValues": [
    "entity-name",
    "7f8d250e-b680-4a50-bfd4-5b5645454523"
  ],
  "filter": {
    "anyOf": [{ "allOf": [{ "key": "kind", "values": ["Documentation"] }] }]
  },
  "firstSortFieldValues": ["another-entity-name", "ad652c98-bf18-4a6b-b848-4545121459f1"],
  "totalItems": 81
}
```

I have observed that this issue occurs only when the last item on the previous page is a particular entity. However, upon thorough examination, I have not been able to identify any issue with this entity.

**Comment by vinzscam** [comment]
> Sure, so I noticed two different issues, **but they seem all happened only when we do cursor request, not the initial request**.
> 
> 1. We get duplicate entities returned from the api, we only have `default` namespace and the duplicated entities are actually the same entityRef.
>    The request of this case:
> 
> ```
> /catalog/entities/by-query?cursor=eyJvcmRlckZpZWxkcyI6W3siZmllbGQiOiJtZXRhZGF0YS5uYW1lIiwib3JkZXIiOiJhc2MifV0sImlzUHJldmlvdXMiOmZhbHNlLCJvcmRlckZpZWxkVmFsdWVzIjpbInBsYXRhZm9ybWEiLCIwZmM0NjU1Yy1jY2M1LTQ3ZGUtYTkxMC03ZTYyYTk1ZTgxNzciXSwiZmlsdGVyIjp7ImFueU9mIjpbeyJhbGxPZiI6W3sia2V5Ijoia2luZCIsInZhbHVlcyI6WyJDb21wb25lbnQiXX1dfV19LCJmaXJzdFNvcnRGaWVsZFZhbHVlcyI6WyJhZ2dyaWdhdG8tc2VydmVyIiwiNTVkNjE2YjQtNzQ5Mi00YmY0LTg1YjUtNzIwMjcyODdiMmQ0Il0sInRvdGFsSXRlbXMiOjY3fQ==&limit=10
> ```
> 
> After decode it is:
> 
> ```json
> {
>   "orderFields": [{ "field": "metadata.name", "order": "asc" }],
>   "isPrevious": false,
>   "orderFieldValues": ["entity-name", "0fc4655c-ccc5-47de-a910-7e45412e8177"],
>   "filter": {
>     "anyOf": [{ "allOf": [{ "key": "kind", "values": ["Component"] }] }]
>   },
>   "firstSortFieldValues": [
>     "another-entity-name",
>     "55d616b4-7492-4bf4-85b5-7512478962d4"
>   ],
>   "totalItems": 67
> }
> ```
> 
> 2. We get entities that are not belong to the Kind we request, for example we get API entities when we request Documentation entities.
>    The request is:
> 
> ```
> /catalog/entities/by-query?cursor=eyJvcmRlckZpZWxkcyI6W3siZmllbGQiOiJtZXRhZGF0YS5uYW1lIiwib3JkZXIiOiJhc2MifV0sImlzUHJldmlvdXMiOmZhbHNlLCJvcmRlckZpZWxkVmFsdWVzIjpbInVwLWxldmVsLW1hbmFnZXItZG9jIiwiN2Y4ZDI1MGUtYjY4MC00YTUwLWJmZDQtNWJkMzhjYjJjMzIzIl0sImZpbHRlciI6eyJhbnlPZiI6W3siYWxsT2YiOlt7ImtleSI6ImtpbmQiLCJ2YWx1ZXMiOlsiRG9jdW1lbnRhdGlvbiJdfV19XX0sImZpcnN0U29ydEZpZWxkVmFsdWVzIjpbImFyZ29jZGRzIiwiYWQ2NTJjOTgtYmYxOC00YTZiLWI4NDgtNDY4NTMzNTUxOWYxIl0sInRvdGFsSXRlbXMiOjgxfQ==&limit=5
> ```
> 
> After decode it is:
> 
> ```json
> {
>   "orderFields": [{ "field": "metadata.name", "order": "asc" }],
>   "isPrevious": false,
>   "orderFieldValues": [
>     "entity-name",
>     "7f8d250e-b680-4a50-bfd4-5b5645454523"
>   ],
>   "filter": {
>     "anyOf": [{ "allOf": [{ "key": "kind", "values": ["Documentation"] }] }]
>   },
>   "firstSortFieldValues": ["another-entity-name", "ad652c98-bf18-4a6b-b848-4545121459f1"],
>   "totalItems": 81
> }
> ```
> 
> **I have observed that this issue occurs only when the last item on the previous page is a particular entity. However, upon thorough examination, I have not been able to identify any issue with this entity.**

Thank you!
I was able to replicate the issue and I've opened another PR https://github.com/backstage/backstage/pull/17539 which adds some tests covering this specific case.
I've changed the target branch of this PR so that we include your commits in #17539

File: .changeset/fifty-grapes-explode.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@backstage/plugin-catalog-backend': patch
+---
+
+Fix a bug in an SQL query where the AND and OR logic is incorrect.
```
Comment (vinzscam) [inline .changeset/fifty-grapes-explode.md:5]:
```suggestion
Fixed a bug that caused filtered entities to be mistakenly included in cursor requests.
```

**Comment by codermango** [comment]
Thank you! Feel free to take over my PR if needed.

**Review by vinzscam — APPROVED** [review]
🚀

**Pull request merged**